pkg/gardenlet/controller/managedseed/managedseed_reconciler.go

```diff
@@ -0,0 +1,280 @@
+	config      *config.GardenletConfiguration
+	imageVector imagevector.ImageVector
+	recorder    record.EventRecorder
+	logger      *logrus.Logger
+}
+
+// newReconciler returns the new ManagedSeed reconciler.
+func newReconciler(ctx context.Context, clientMap clientmap.ClientMap, config *config.GardenletConfiguration, imageVector imagevector.ImageVector, recorder record.EventRecorder, logger *logrus.Logger) reconcile.Reconciler {
+	return &reconciler{
+		ctx:         ctx,
+		clientMap:   clientMap,
+		config:      config,
+		imageVector: imageVector,
+		recorder:    recorder,
+		logger:      logger,
+	}
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	gardenClient, err := r.clientMap.GetClient(r.ctx, keys.ForGarden())
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get garden client: %w", err)
+	}
+
+	managedSeed := &seedmanagementv1alpha1.ManagedSeed{}
+	if err := gardenClient.Client().Get(r.ctx, request.NamespacedName, managedSeed); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Debugf("Skipping ManagedSeed %s because it has been deleted", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		r.logger.Errorf("Could not get ManagedSeed %s from store: %v", request.NamespacedName, err)
+		return reconcile.Result{}, err
+	}
+
+	if managedSeed.Spec.Shoot == nil || managedSeed.Spec.Shoot.Name == "" {
+		r.logger.Errorf("Skipping ManagedSeed %s because it does not specify a shoot", request.NamespacedName)
+		return reconcile.Result{}, nil
+	}
+
+	if managedSeed.DeletionTimestamp != nil {
+		return r.delete(gardenClient, managedSeed)
+	}
+	return r.createOrUpdate(gardenClient, managedSeed)
+}
+
+func (r *reconciler) createOrUpdate(gardenClient kubernetes.Interface, managedSeed *seedmanagementv1alpha1.ManagedSeed) (reconcile.Result, error) {
+	managedSeedLogger := logger.NewFieldLogger(r.logger, "managedSeed", kutil.ObjectName(managedSeed))
+
+	// Ensure gardener finalizer
+	if err := controllerutils.EnsureFinalizer(r.ctx, gardenClient.DirectClient(), managedSeed, gardencorev1beta1.GardenerName); err != nil {
+		managedSeedLogger.Errorf("Could not ensure gardener finalizer on ManagedSeed: %+v", err)
+		return reconcile.Result{}, err
+	}
+
+	// Get shoot
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := gardenClient.Client().Get(r.ctx, kutil.Key(managedSeed.Namespace, managedSeed.Spec.Shoot.Name), shoot); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if the shoot is reconciled
+	if shoot.Generation != shoot.Status.ObservedGeneration || shoot.Status.LastOperation == nil || shoot.Status.LastOperation.State != gardencorev1beta1.LastOperationStateSucceeded {
+		managedSeedLogger.Infof("Waiting for shoot %s to be reconciled before registering it as seed", kutil.ObjectName(shoot))
+
+		// Update status to Pending
+		if updateErr := updateStatusPending(r.ctx, gardenClient.DirectClient(), managedSeed, "Reconciliation pending"); updateErr != nil {
```

Comment:
Can you switch to patch requests for updating the status and get rid of the `DirectClient` usage, please?

Reply:
OK